python/cpython
# bpo-44934: Add optional feature AppendPath to Windows msi installer

### Initial problem
Automatically adding Python to the `PATH` is convenient for Windows end users; this was achieved by the installer by prepending the Python Install and Scripts dir to `PATH`. However this might cause conflicts with system executables.

How to reproduce:

1. Install Python msi system wide 
2. Choose `Add Python to environment variables` optional features
3. Install [impacket](https://github.com/SecureAuthCorp/impacket) in a system context (e.g. administrativ shell, `pip install impacket`

Installing impacket in a system context drops - among others - the following Python scripts to the Scripts dir:

- `reg.py`
- `ping.py`

Some batch scripts using commands like

- `REG ADD HKEY_CURRENT_USER\Console /v Test /d "Test Data" `
- `ping domain.controller`

that expect to run `C:\Windows\System32\reg.exe` now run `C:\Program Files\Python\Scripts\reg.py`.

### Conflicts with other use cases

After reporting my problem @zooba reasoned that Python developers might have different versions of Python installed and to make sure they always use the version installed last he chose to prepend those directories instead of appending them to `PATH`.

### Proposed solution

In order to provide system administrators with a way to automatically add Python to `PATH` by appending it, @zooba suggested to create an command line only separate optional feature for this. Therefore - without any preexisting knowledge of msi or the WiX toolset - I browsed the existing code for the path optional feature and basically duplicated the occurences of that optional feature.

### Possible problems

I did not introduce any layout in the GUI of the installer. I also didn't implement any checks if e.g. the `path` and `appendpath` optional feature are set simultaneousely when running the installer via CLI, thus a user might have 2 occurences of the install and scripts directory if they chosse to use both flags. @zooba said in the BPO report that this would probably be fine.

Also my PR might be very far from being able to succesfully compile since I have no prior experience of thw WiX toolset and couldn't get a clean build with the provided instructions in [Tools/msi/README.txt](https://github.com/python/cpython/blob/main/Tools/msi/README.txt) on one of the free Windows 10 Edge Test VM even before modifying the source code. @zooba suggested to submit this PR anyway since the build pipeline would show if everything builds.





https://bugs.python.org/issue44934
